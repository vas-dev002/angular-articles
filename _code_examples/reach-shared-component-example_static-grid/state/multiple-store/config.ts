import { InjectionToken, Type } from "@angular/core";
import { ActionReducer } from "@ngrx/store";

export const MULTIPLE_STORE_REDUCER = new InjectionToken<ActionReducer<any, any>>("MultipleStoreReducer");
export const MULTIPLE_STORE_SLICE_NAME = new InjectionToken<string>("StoreSliceName");
export const MULTIPLE_STORE_EFFECTS = new InjectionToken<Type<unknown>>("MultipleStoreEffects");
