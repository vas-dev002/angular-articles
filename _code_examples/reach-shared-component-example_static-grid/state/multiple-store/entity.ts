export interface MultipleStoreMeta {
	sliceName: string;
}
