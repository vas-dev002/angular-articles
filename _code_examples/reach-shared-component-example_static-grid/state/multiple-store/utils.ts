import { Action } from "@ngrx/store";
import { filter } from "rxjs/operators";
import { isEmptyNullOrUndefined } from "das-lib/utils-by-type";
import { MultipleStoreMeta } from "./entity";

export function createLocalAction<A extends Action>(action: A, name: string): any {
	if(!isEmptyNullOrUndefined(name)) {
		(action as any).__multipleStoreMeta = { sliceName: name } as MultipleStoreMeta;
	}
	return action;
}

export function isComponentAction(meta: MultipleStoreMeta, name: string): boolean {
	return meta.sliceName === name;
}

export function forThisComponent(name: string) {
	return filter(
		(action: any) => !action.__multipleStoreMeta || isComponentAction(action.__multipleStoreMeta, name)
	);
}
