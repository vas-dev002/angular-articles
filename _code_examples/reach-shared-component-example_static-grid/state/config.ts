import { Provider } from "@angular/core";

import { StaticGridEffects } from "./effects/effect";
import { multipleStoreProviders } from "./multiple-store/multiple-store";
import { staticGridReducer } from "./reducers/reducer";
import { GridSelectors } from "./selectors/selector";
import { TopBarSelectors } from "./selectors/topbar";

export function multipleGridStoreProviders(gridKey: string): Provider[] {
	return multipleStoreProviders(
		gridKey,
		staticGridReducer,
		StaticGridEffects,
		[GridSelectors, TopBarSelectors]
	);
}
