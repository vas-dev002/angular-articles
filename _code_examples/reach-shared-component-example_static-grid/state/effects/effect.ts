import { Inject, Injectable, Optional } from "@angular/core";
import { createEffect, ofType, Actions } from "@ngrx/effects";
import * as act from "../actions/action";
import {
	map,
	mergeMap,
	switchMap,
	catchError,
	onErrorResumeNext,
	takeUntil
} from "rxjs/operators";
import { ApiService } from "../../../../../../components/services/api";
import { LayoutErrorMessage } from "../../../../../../state/layout/actions";
import { throwError, forkJoin } from "rxjs";
import { apiError } from "../../../../../../components/util/common/error";
import * as actTopbar from "../../state/actions/topbar";
import { IGroupByColumn } from "../../components/topbar/entity";
import { Action } from "@ngrx/store";
import { IDataRowSG, IColumnMetaSG } from "../../components/main/entity";
import { IStaticGridGetDataParams, IStaticGridGetMetadataParams } from "../../../util/entity";
import { objectByPropertyStringComparator } from "../../../../../evergreen-shared/utils/comparators";
import { createLocalAction, forThisComponent } from "../multiple-store/utils";
import { MULTIPLE_STORE_SLICE_NAME } from "../multiple-store/config";
import { MultipleStoreEffects } from "../multiple-store/multiple-store.effects";
import { ITypedController } from "../../components/columns/entity";

@Injectable()
export class StaticGridEffects extends MultipleStoreEffects {

	constructor(
		@Optional() @Inject(MULTIPLE_STORE_SLICE_NAME) protected sliceName: string,
		protected actions$: Actions,
		private readonly api: ApiService
	) {
		super(actions$, sliceName);
	}

	public readonly staticGridDataGetRequest$ = createEffect(() =>
		this.actions$.pipe(
			ofType(act.StaticGridDataRequest),
			forThisComponent(this.sliceName),
			map(
				(prm: {
					action: Action,
					url: string;
					requestTimeout?: number;
					params: string[];
					transformMeta: (response: Object, customMetaTypes: ITypedController[]) => IColumnMetaSG[];
					customMetaTypes?: ITypedController[];
					transformData: (response: Object) => IDataRowSG[];
					errorActions: (res: any) => Action[];
				}) => prm
			),
			switchMap(prm => {
				return this.api.getRequest(prm.url, prm.params, prm.requestTimeout).pipe(
					takeUntil(
						this.actions$.pipe(
								ofType(act.CancelStaticGridDataRequest)
						)),
					map((response: IDataRowSG[]) => this.mapRowsData(response, prm)),
					catchError(res => throwError({ res, prm })),
					mergeMap(result => this.mapStaticGridData(result)),
					catchError(err => this.errorHandler(err)),
					onErrorResumeNext()
				);
			}

			),
			map(prm => prm)
		)
	);

	public readonly staticGridSeparatedDataRequest$ = createEffect(() =>
		this.actions$.pipe(
			ofType(act.StaticGridSeparatedDataRequest),
			forThisComponent(this.sliceName),
			map((prm: { payload: IStaticGridGetDataParams }) => prm),
			switchMap((params: { payload: IStaticGridGetDataParams }) => {
				const getMetadataUrl = params.payload.urlMetadata ? params.payload.urlMetadata : `${params.payload.url}/metadata`;

				return forkJoin([
					this.api.getRequest(params.payload.url, params.payload.params),
					this.api.getRequest(getMetadataUrl, params.payload.params)
				]).pipe(
					takeUntil(
						this.actions$.pipe(ofType(act.CancelStaticGridDataRequest))
					),
					map(([results, metadata]) =>
						({ results: this.sortDataByColumnName(results, params.payload.sortFieldName), metadata })),
					map((response: any) => this.mapRowsData(response, params.payload)),
					catchError(res => throwError({ res, params })),
					mergeMap(result => this.mapStaticGridData(result)),
					catchError(err => this.errorHandler(err)),
					onErrorResumeNext()
				);
			}
			)
		)
	);

	public readonly staticGridMetadataRequest$ = createEffect(() =>
		this.actions$.pipe(
			ofType(act.StaticGridMetadataRequest),
			forThisComponent(this.sliceName),
			map((prm: { payload: IStaticGridGetMetadataParams }) => prm),
			switchMap((params: { payload: IStaticGridGetMetadataParams }) => {

				return this.api.getRequest(params.payload.url, params.payload.params).pipe(
					takeUntil(
						this.actions$.pipe(ofType(act.CancelStaticGridDataRequest))
					),
					map((metadata) => ({ res: { meta: params.payload.transformMeta({ metadata }) } })),
					catchError(res => throwError({ res, prm: params.payload })),
					mergeMap(result => this.mapStaticGridData(result)),
					catchError(err => this.errorHandler(err)),
					onErrorResumeNext()
				);
			}
			)
		)
	);

	private sortDataByColumnName(data: any, fieldName: string): any {
		return fieldName ? [...data.sort(objectByPropertyStringComparator(fieldName))] : data;
	}

	private errorHandler(err) {
		const errActions: Action[] = err.prm.errorActions(err.res);
		const error = apiError(err, "FullPage");
		let resultErrorActions = [];
		if (errActions && errActions.length > 0) {
			resultErrorActions = [...errActions];
		} else {
			resultErrorActions = [LayoutErrorMessage({ message: error })];
		}
		resultErrorActions.push(createLocalAction(act.StaticGridLoadError(), this.sliceName));
		return resultErrorActions;
	}

	private mapRowsData(response: IDataRowSG[], prm) {
		const res = {
			meta: prm.transformMeta(response, prm.customMetaTypes),
			data: prm.transformData(response)
		};
		return { res, prm };
	}

	private mapStaticGridData(result, onlyMetadata = false) {
		const res = result.res;
		let actions = [
			createLocalAction(act.StaticGridSelectRows({ rows: []}), this.sliceName),
			createLocalAction(act.StaticGridSetMeta({ meta: res.meta }), this.sliceName),
			createLocalAction(act.StaticGridDataLoaded({ data: res.data }), this.sliceName),
			createLocalAction(actTopbar.TopBarSetGroupColumns({
				GroupColumns: res.meta
					.filter(x => (this.isColumnGroupable(x)))
					.map(x => {
						return {
							field: x.columnName,
							name: x.headerName
						} as IGroupByColumn;
					})
			}), this.sliceName)
		];

		if (onlyMetadata) {
			actions = actions.splice(2, 1);
		}
		return actions;
	}

	private isColumnGroupable(columnMeta: IColumnMetaSG): boolean {
		return columnMeta.visible;
	}
}
