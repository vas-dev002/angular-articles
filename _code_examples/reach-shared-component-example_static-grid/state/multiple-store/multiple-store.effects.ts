import { Injectable } from "@angular/core";
import { ofType, Actions, OnIdentifyEffects, OnRunEffects, EffectNotification } from "@ngrx/effects";
import { takeUntil } from "rxjs/operators";
import { Observable } from "rxjs";
import { forThisComponent } from "../multiple-store/utils";

@Injectable()
export abstract class MultipleStoreEffects implements OnIdentifyEffects, OnRunEffects {

	constructor(protected readonly actions$: Actions, protected readonly sliceName: string) {}

	ngrxOnIdentifyEffects(): string {
		return this.sliceName;
	}

	ngrxOnRunEffects(resolvedEffects$: Observable<EffectNotification>): Observable<EffectNotification> {
		return resolvedEffects$.pipe(
			takeUntil(
				this.actions$.pipe(
					ofType(`[${this.sliceName}] Destroyed`),
					forThisComponent(this.sliceName)
				)
			)
		);
	}
}
