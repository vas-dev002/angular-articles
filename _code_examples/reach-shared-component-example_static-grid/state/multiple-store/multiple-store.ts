import { Inject, Injectable, OnDestroy, Provider, Type } from "@angular/core";
import {
	ReducerManager,
	ActionReducer,
	Store,
	Action
} from "@ngrx/store";
import { EffectSources } from "@ngrx/effects";
import { createLocalAction, isComponentAction } from "./utils";
import { MULTIPLE_STORE_EFFECTS, MULTIPLE_STORE_REDUCER, MULTIPLE_STORE_SLICE_NAME } from "./config";

@Injectable()
export class MultipleStore implements OnDestroy {
	name = "";

	initActionType = "";
	destroyedActionType = "";
	setValueActionType = "";

	constructor(
		private readonly store: Store,
		private readonly reducers: ReducerManager,
		private readonly effectsSources: EffectSources,
		@Inject(MULTIPLE_STORE_REDUCER) readonly reducer: ActionReducer<any, any>,
		@Inject(MULTIPLE_STORE_EFFECTS) readonly effects: Type<unknown>,
		@Inject(MULTIPLE_STORE_SLICE_NAME) readonly storeSliceName: string = null
	) {
		this.name = storeSliceName || `${this.constructor.name}`;
		this.initActionType = `[${this.name}] Init`;
		this.destroyedActionType = `[${this.name}] Destroyed`;
		this.setValueActionType = `[${this.name}] Set Value`;

		this.reducers.addReducer(this.name, this.reducerProxy );

		this.dispatch({type: this.initActionType});
		this.effectsSources.addEffects(effects);
	}

	reducerProxy = (state, action) => {
		// global actions pass through, invoke the component reducer
		if (!action.__multipleStoreMeta) {
			return this.reducer(state, action);
		}

		// filter out local actions for this component
		if (!isComponentAction(action.__multipleStoreMeta, this.name)) {
			return state;
		}

		// update state
		if (action.type === this.setValueActionType) {
			return action.value;
		}

		// local action, invoke the component reducer
		return this.reducer(state, action);
	}

	dispatch<A extends Action>(action: A): void {
		this.store.dispatch(createLocalAction(action, this.name));
	}

	ngOnDestroy(): void {
		this.reducers.removeReducer(this.name);
		this.dispatch({ type: this.destroyedActionType});
	}
}

export function multipleStoreProviders(
	sliceName: string,
	reducer: ActionReducer<any, any>,
	effects: Type<unknown>,
	selectors: Type<unknown>[]
): Provider[] {
	return [
		{
			provide: MULTIPLE_STORE_SLICE_NAME,
			useValue: sliceName
		},
		MultipleStore,
		{
			provide: MULTIPLE_STORE_REDUCER,
			useValue: reducer
		},
		{
			provide: MULTIPLE_STORE_EFFECTS,
			useClass: effects
		},
		...selectors
	];
}
