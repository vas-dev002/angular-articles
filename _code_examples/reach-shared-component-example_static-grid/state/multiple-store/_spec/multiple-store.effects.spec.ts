import { Optional } from "@ag-grid-community/core";
import { Inject } from "@angular/core";
import { TestBed } from "@angular/core/testing";
import { Actions } from "@ngrx/effects";
import { provideMockActions } from "@ngrx/effects/testing";
import { Action } from "@ngrx/store";
import { Observable, of } from "rxjs";
import { MULTIPLE_STORE_SLICE_NAME } from "../config";
import { MultipleStoreEffects } from "../multiple-store.effects";
declare const expect: jest.Expect;

class TestMultipleStoreEffects extends MultipleStoreEffects {
	constructor(protected readonly actions$: Actions,
		@Optional() @Inject(MULTIPLE_STORE_SLICE_NAME) protected sliceName: string) {
		super(actions$, sliceName);
	}
}

describe("Multiple Store Effects", () => {
	let effects: TestMultipleStoreEffects;
	let actions: Observable<Action>;
	const sliceName = "TestKey";

	beforeEach(() => {
		TestBed.configureTestingModule({
			providers: [
				{
					provide: MULTIPLE_STORE_SLICE_NAME,
					useValue: sliceName
				},
				TestMultipleStoreEffects,
				provideMockActions(() => actions)
			]
		});
		actions = TestBed.inject(Actions);
		effects = TestBed.inject(TestMultipleStoreEffects);

	});

	it("should be created", () => {
		expect(effects).toBeTruthy();
		expect(actions).toBeTruthy();
	});

	it("ngrxOnIdentifyEffects - should return slice name", () => {
		const name = effects.ngrxOnIdentifyEffects();
		expect(name).toEqual(sliceName);
	});

	it("ngrxOnRunEffects - should return observable", () => {
		const res = effects.ngrxOnRunEffects(of());
		expect(res).toMatchSnapshot();
	});
});
