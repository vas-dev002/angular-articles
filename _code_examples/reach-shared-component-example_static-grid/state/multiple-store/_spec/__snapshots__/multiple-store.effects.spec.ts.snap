// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple Store Effects ngrxOnRunEffects - should return observable 1`] = `
Observable {
  "_isScalar": false,
  "operator": TakeUntilOperator {
    "notifier": Actions {
      "_isScalar": false,
      "operator": FilterOperator {
        "predicate": [Function],
        "thisArg": undefined,
      },
      "source": Actions {
        "_isScalar": false,
        "operator": FilterOperator {
          "predicate": [Function],
          "thisArg": undefined,
        },
        "source": Actions {
          "_isScalar": false,
          "source": Observable {
            "_isScalar": false,
            "_subscribe": [Function],
          },
        },
      },
    },
  },
  "source": Observable {
    "_isScalar": false,
    "_subscribe": [Function],
  },
}
`;
